### PR TITLE
[Enhancement]  phase 2: split implementation of SenderQueue for non-pipeline and pipeline

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [x] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

see more details here: https://github.com/StarRocks/starrocks/pull/6766

Major changes in this PR:
1. split the implementation of SenderQueue for pipeline and non-pipeline
2. the non-pipeline implementation is almost the same as before, without lots of changes
3. the pipeline implementation introduces a lock-free queue and some atomic variables to avoid lock competition
4. there is a lot of redundant codes in add_chunks and add_chunks_and_keep_order, I have simplified these codes by adding a template function in both NonPipelineSenderQueue and PipelineSenderQueue

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: be/src/runtime/current_thread.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 #include <string>
 
+#include "fmt/format.h"
```
Comment (kangkaisen) [inline be/src/runtime/current_thread.h:7]: why need this include?

File: be/src/runtime/sender_queue.h [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    Status _build_chunk_meta(const ChunkPB& pb_chunk);
+
+    Status _deserialize_chunk(const ChunkPB& pchunk, vectorized::Chunk* chunk, faststring* uncompressed_buffer);
+
+    DataStreamRecvr* _recvr = nullptr;
+
+    serde::ProtobufChunkMeta _chunk_meta;
+};
+
+class DataStreamRecvr::NonPipelineSenderQueue : public DataStreamRecvr::SenderQueue {
```
Comment (kangkaisen) [inline be/src/runtime/sender_queue.h:56]: mark final

File: be/src/runtime/sender_queue.h [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        // When the memory of the ChunkQueue exceeds the limit,
+        // we have to hold closure of the request, so as not to let the sender continue to send data.
+        // A Request may have multiple Chunks, so only when the last Chunk of the Request is consumed,
+        // the callback is closed->run() Let the sender continue to send data
+        google::protobuf::Closure* closure = nullptr;
+        ChunkUniquePtr chunk_ptr;
+        // Time in nano of saving closure
+        int64_t queue_enter_time;
+
+        inline static ChunkItem create(int64_t chunk_bytes, int32_t driver_sequence, google::protobuf::Closure* closure,
```
Comment (kangkaisen) [inline be/src/runtime/sender_queue.h:170]: Why need a static method? not by construct.

File: be/src/runtime/sender_queue.cpp [hunk 4]
```diff
@@ -0,0 +1,867 @@
...
+            auto& pchunk = request.chunks(0);
+            RETURN_IF_ERROR(_build_chunk_meta(pchunk));
+        }
+    }
+
+    size_t total_chunk_bytes = 0;
+    faststring uncompressed_buffer;
+    ChunkQueue local_chunk_queue;
+
+    if (use_pass_through) {
```
Comment (kangkaisen) [inline be/src/runtime/sender_queue.cpp:345]: Only pipeline engine support use_pass_through

File: be/src/runtime/sender_queue.cpp [hunk 5]
```diff
@@ -0,0 +1,867 @@
...
+        int index = _is_pipeline_level_shuffle ? chunk.driver_sequence : 0;
+        size_t chunk_bytes = chunk.chunk_bytes;
+        _chunk_queues[index].enqueue(std::move(chunk));
+        _total_chunks++;
+        _recvr->_num_buffered_bytes += chunk_bytes;
+    }
+    return Status::OK();
+}
+
+Status DataStreamRecvr::PipelineSenderQueue::add_chunks_and_keep_order(const PTransmitChunkParams& request,
```
Comment (kangkaisen) [inline be/src/runtime/sender_queue.cpp:666]: add_chunks has redundant codes with add_chunks_and_keep_order method

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run all

**Comment by silverbullet233** [comment]
run all

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: be/src/runtime/current_thread.h (hunk 1 above)
Comment (silverbullet233) [reply be/src/runtime/current_thread.h:7]: > why need this include?

The relevant interaface is called in this header file. If it is not explicitly included, the compilation will fail.

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: be/src/runtime/sender_queue.cpp (hunk 5 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp]: > add_chunks has redundant codes with add_chunks_and_keep_order method

fixed by adding a template function

**Review by kangkaisen — DISMISSED** [review]

File: be/src/runtime/sender_queue.cpp [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+        return;
+    }
+    _sender_eos_set.insert(be_number);
+    DCHECK_GT(_num_remaining_senders, 0);
+    _num_remaining_senders--;
+    VLOG_FILE << "decremented senders: fragment_instance_id=" << print_id(_recvr->fragment_instance_id())
+              << " node_id=" << _recvr->dest_node_id() << " #senders=" << _num_remaining_senders
+              << " be_number=" << be_number;
+    if (_num_remaining_senders == 0) {
+        _data_arrival_cv.notify_one();
```
Comment (liuyehcf) [inline be/src/runtime/sender_queue.cpp:337]: Why notify_one not notify_all

File: be/src/runtime/sender_queue.cpp [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+        _max_processed_sequences.lazy_emplace(be_number, [be_number](const auto& ctor) { ctor(be_number, -1); });
+
+        _buffered_chunk_queues.lazy_emplace(be_number, [be_number](const auto& ctor) {
+            ctor(be_number, phmap::flat_hash_map<int64_t, ChunkList>());
+        });
+
+        auto& chunk_queues = _buffered_chunk_queues[be_number];
+
+        if (!chunks.empty() && done != nullptr && _recvr->exceeds_limit(total_chunk_bytes)) {
+            chunks.back().closure = *done;
```
Comment (liuyehcf) [inline be/src/runtime/sender_queue.cpp:618]: Missing update queue_enter_time?

Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp:618]: > Missing update `queue_enter_time`?

fixed

File: be/src/runtime/sender_queue.cpp (hunk 6 above)
Comment (silverbullet233) [reply be/src/runtime/sender_queue.cpp]: > Why notify_one not notify_all

this is the original logic, I think notify_all makes more sense, I have fixed it.

**Comment by silverbullet233** [comment]
run all

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by liuyehcf — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Pull request merged**